### PR TITLE
zset: add in-place fast path for score updates in listpack encoding

### PR DESCRIPTION
**Pull request by charsyam**
Summary

- Before: Updating the score of an existing member in a listpack-encoded zset always performs zzlDelete followed by zzlInsert, which re-scans
from the head to find the new position.
- Change: If the new score keeps the member between its current neighbors under (score, ele) ordering, update the score entry in place via
lpReplaceInteger or lpReplace—no delete, no scan.
- Note: Mirrors the fast path already used by zslUpdateScore on the skiplist side.

Motivation

- Small score adjustments are common (leaderboards, counters, weighted metrics). In these patterns, the member typically stays in place, yet the
current path still pays for delete + O(N/2) re-scan. Avoiding that work makes the hot path much cheaper.

Change

- Two commits:
    - add in-place fast path: when score != curscore, probe neighbors; if ordering holds, replace the score entry with lpReplace*; otherwise fall
back to delete + re-insert.
    - probe only one neighbor: only the neighbor in the direction of the score change can violate ordering.
    - If `score > old`, the previous neighbor is always safe; check only the next.
    - If `score < old`, the next is always safe; check only the previous.
- No protocol, syntax, error, or semantic change. The slow path is unchanged.

Benchmarks

- Setup: 20k ops per run, 5 runs per cell, μs/op (mean ± σ). zset-max-listpack-entries raised to keep listpack encoding. RESP/loopback cost
included.
- small_delta (±0.5): fast path almost always hits.
    - N=128: 3.195 → 1.457 μs/op (~2.19×)
    - N=512: 8.188 → 2.406 μs/op (~3.40×)
- monotonic (increasing, ZINCRBY-like): consistent small gains.
    - N=512: 13.196 → 11.923 μs/op (~1.11×)
- random (adversarial): fast path rarely hits; results ~equal to baseline, variation within ~1σ.
    - N=32: +3.0% (~1σ), others near tie

Safety

- Behaviorally identical: same replies, errors, and tie-break (score, then lex).
- Reallocation-safe: lpReplace* may reallocate; the returned pointer is always written back via objectSetVal.
- Floating-point edge cases: NaN is rejected at zsetAdd entry; +0.0/-0.0 compare equal so the fast path is not taken; ±Inf follows the existing
d2string/double2ll encoding paths.
- Fallback remains: if ordering doesn’t hold, we still do zzlDelete → zzlInsert.

Tests

- Existing zset tests pass. Updates of existing scores (ZADD overwrite, ZINCRBY, ZADD GT/LT) naturally exercise the fast path; reordering cases
exercise the slow path.

## Benchmarks

  Linux, 20k ops per run, 5 runs per cell, mean ± σ (μs/op). `zset-max-listpack-entries` raised to keep listpack encoding. RESP and loopback cost
   included.

  ### small_delta (±0.5 nudge — fast path hits almost always)

  | N   | baseline        | feature         | speedup   |
  |---:|---:|---:|---:|
  |   8 | 1.371 ± 0.030   | 1.214 ± 0.005   | **1.13×** |
  |  16 | 1.476 ± 0.035   | 1.281 ± 0.046   | **1.15×** |
  |  32 | 1.657 ± 0.017   | 1.276 ± 0.023   | **1.30×** |
  |  64 | 2.094 ± 0.073   | 1.340 ± 0.016   | **1.56×** |
  | 128 | 3.195 ± 0.042   | 1.457 ± 0.019   | **2.19×** |
  | 256 | 4.797 ± 0.024   | 1.727 ± 0.021   | **2.78×** |
  | 512 | 8.188 ± 0.084   | 2.406 ± 0.024   | **3.40×** |

  ### monotonic (increasing scores — ZINCRBY-like)

  | N   | baseline        | feature         | speedup |
  |---:|---:|---:|---:|
  |   8 | 1.439 ± 0.018   | 1.439 ± 0.057   | tie     |
  |  16 | 1.685 ± 0.051   | 1.609 ± 0.039   | ~1.05×  |
  |  32 | 2.024 ± 0.065   | 1.978 ± 0.123   | ~1.02×  |
  |  64 | 2.955 ± 0.168   | 2.713 ± 0.148   | ~1.09×  |
  | 128 | 4.634 ± 0.098   | 4.534 ± 0.086   | ~1.02×  |
  | 256 | 7.386 ± 0.145   | 6.927 ± 0.136   | ~1.07×  |
  | 512 | 13.196 ± 0.131  | 11.923 ± 0.183  | **~1.11×** |

  ### random (adversarial — fast path rarely hits)

  | N   | baseline        | feature         | Δ            |
  |---:|---:|---:|---:|
  |   8 | 1.407 ± 0.048   | 1.388 ± 0.010   | −1.4%        |
  |  16 | 1.486 ± 0.044   | 1.484 ± 0.022   | tie          |
  |  32 | 1.653 ± 0.032   | 1.702 ± 0.052   | +3.0% (~1σ)  |
  |  64 | 2.118 ± 0.073   | 2.120 ± 0.038   | tie          |
  | 128 | 3.242 ± 0.067   | 3.229 ± 0.036   | −0.4%        |
  | 256 | 4.808 ± 0.009   | 4.861 ± 0.050   | +1.1% (~1σ)  |
  | 512 | 8.225 ± 0.047   | 8.291 ± 0.052   | +0.8% (~1σ)  |

### CODE REVIEW

File: src/t_zset.c [hunk 1]
```diff
@@ -1507,8 +1507,56 @@ int zsetAdd(robj *zobj, double score, sds ele, int in_flags, int *out_flags, dou
...
+                /* Fast path: if the new score keeps the element in its current
+                 * slot under (score, ele) ordering, replace the score entry in
+                 * place instead of delete + full re-scan.
+                 *
+                 * Only the neighbor in the direction of the score change needs
+                 * to be checked: if the score increases, the previous neighbor
+                 * satisfies (prev_score < score) strictly (since prev_score <=
+                 * curscore < score), so only the next neighbor can violate the
+                 * order. The score-decrease case is symmetric. */
+                int keep_position = 1;
```
Comment (rainsupreme) [inline src/t_zset.c]: nit: a `bool` would be better for keep_position

**Review by rainsupreme — APPROVED** [review]
This looks like a very reasonable improvement! I only had one minor nit complaint. I appreciate the benchmarking too and that you included the error margins - that helps a lot. :)

File: src/t_zset.c (hunk 1 above)
Comment (daemyung-lablup) [reply src/t_zset.c]: @rainsupreme Thanks. applied. :)

Comment (charsyam) [reply src/t_zset.c]: @rainsupreme Is there any comments or progress?

Comment (rainsupreme) [reply src/t_zset.c]: @charsyam The next step is to get a maintainer or committer to sign off and merge it. That's these people: https://github.com/valkey-io/valkey/blob/unstable/MAINTAINERS.md

I requested a review from @ranshid, though of course any of them can do the review. If nobody's reviewed in in a week or so then request someone else or maybe ask in slack. The maintainers are busy people and there's a large PR backlog, after all. 🙂

File: src/t_zset.c [hunk 2]
```diff
@@ -1507,8 +1507,56 @@ int zsetAdd(robj *zobj, double score, sds ele, int in_flags, int *out_flags, dou
 
             /* Remove and re-insert when score changed. */
             if (score != curscore) {
-                objectSetVal(zobj, zzlDelete(objectGetVal(zobj), eptr));
-                objectSetVal(zobj, zzlInsert(objectGetVal(zobj), ele, score));
+                unsigned char *zl = objectGetVal(zobj);
```
Comment (ranshid) [inline src/t_zset.c]: I think it worth wrapping this logic in zzlUpdateScore function which will make this code more aligned with the way we currently use zslUpdateScore. 
Then the implementation here would simplycollapses to:

```
-            /* Remove and re-insert when score changed. */
             if (score != curscore) {
-                objectSetVal(zobj, zzlDelete(objectGetVal(zobj), eptr));
-                objectSetVal(zobj, zzlInsert(objectGetVal(zobj), ele, score));
+                objectSetVal(zobj,
+                             zzlUpdateScore(objectGetVal(zobj), eptr, ele, score, curscore));
                 *out_flags |= ZADD_OUT_UPDATED;
             }
```

**Review by ranshid — COMMENTED** [review]
I am sorry for missing this! (probably due to local holidays)

I think it generally looks fine. I would consider to improve the maintainability and symethry with zsl by introducing a zzlUpdateScore variant (see inline comment).

Note regarding the tests:

The PR claims existing tests cover this ("ZADD overwrite, ZINCRBY, ZADD GT/LT"). I think that MIGHT be partially true — they exercise both the new fast path and the old slow path by luck of input. But the fast-patt/slow-
path boundary has subtle invariants that existing tests weren't designed to target, and a bug here silently corrupts listpack ordering without immediate crash. I'd like to see explicit listpack-encoded 
cases for:

1. Head element — no previous neighbor (score decrease case).
2. Tail element — no next neighbor (score increase case).
3. Tie score with lex ordering at the boundary — update to a score equal to the neighbor's, where lex comparison is the deciding factor for keep vs move. Both directions.
4. Integer-encoded ↔ string-encoded score transition — e.g., old score 42 (integer), new score 42.5 (string), same position. Verifies lpReplace path isn't conflated with lpReplaceInteger.
5. Cross-boundary update — score change that moves the element exactly one slot (hits slow path via neighbor-violates).
6. Single-element zset — trivial fast path, no neighbors at all.

```
### tests/unit/type/zset.tcl — suggested addition

test "ZADD in-place update: tie score with lex ordering - $encoding" {
    r del ztmp
    r zadd ztmp 1 a 2 b 2 c 3 d
    # Update 'a' to score 2 — ties with 'b' and 'c'; lex "a" < "b" < "c",
    # so 'a' should come before 'b' in the (score=2) range.
    r zadd ztmp 2 a
    assert_equal {b c a d} [r zrange ztmp 0 -1]
    # Wait, actual expected: {b c a d} if reordering, but since score 2 and a<b lex,
    # it should be {a b c d}. Verify fast-path/slow-path chose correctly.
}

test "ZADD in-place update: integer-to-string score encoding - $encoding" {
    r del ztmp
    r zadd ztmp 10 a
    r zadd ztmp 10.5 a        ;# integer → string, same position
    assert_equal 10.5 [r zscore ztmp a]
}

test "ZADD in-place update: head and tail edge cases - $encoding" {
    r del ztmp
    r zadd ztmp 1 a 2 b 3 c
    r zadd ztmp 0.5 a         ;# head, decrease: no prev neighbor
    r zadd ztmp 4 c           ;# tail, increase: no next neighbor
    assert_equal {a b c} [r zrange ztmp 0 -1]
}
```

**Review by ranshid — APPROVED** [review]
Thank you! looks good

@charsyam - can you please fix the DCO missing signoff?

**Comment by charsyam** [comment]
@ranshid Thanks. I signed off.

**Review by sarthakaggarwal97 — APPROVED** [review]
Looks good to me!

**Comment by sarthakaggarwal97** [comment]
@ranshid should we merge this?
